include/MatrixClient.h

```diff
@@ -52,9 +52,9 @@ class MatrixClient : public QNetworkAccessManager
         void downloadImage(const QString &event_id, const QUrl &url);
         void downloadFile(const QString &event_id, const QUrl &url);
         void messages(const QString &room_id, const QString &from_token, int limit = 30) noexcept;
-        void uploadImage(const QString &roomid, const QString &filename);
-        void uploadFile(const QString &roomid, const QString &filename);
-        void uploadAudio(const QString &roomid, const QString &filename);
+        void uploadImage(const QString &roomid, QSharedPointer<QIODevice> iodev);
+        void uploadFile(const QString &roomid, QSharedPointer<QIODevice> iodev);
+        void uploadAudio(const QString &roomid, QSharedPointer<QIODevice> iodev);
```

Comment:
Why do these need to be shared pointers?

Reply:
They're shared pointers because notice the lambda in the `MatrixClient::upload*()` member functions for responding to a `QNetworkReply::finished` signal. By the time that lambda is executed (asynchronously), we have lost the scope of the `QIODevice`. If I were to pass by const ref, the reference would be bound to a temporary local, which would cause a segfault somewhere during the execution of the lambda. To resolve this, the best method I could think of was to create a pointer so that it's allocated on the heap. 

I would love to hear if you have a better method.